### DISCUSSION TITLE
Transparent proxy with redirector: What is the virtual device for?

### DISCUSSION THREAD

**Discussion by da-kami** [mitmproxy/mitmproxy#7171] [Q&A]
Hi mitmproxy community!

I have been analyzing the mitmproxy Rust (and python) code to get a better understanding how the transparent on-device setup using a network-level redirector works.

I have been mostly focusing on the TCP flow on Windows within https://github.com/mitmproxy/mitmproxy_rs 

As far as I understand your solution involves a virtual device that simulates the remote server for the application, i.e. all connections will go through the virtual device which will establish the connection to an application that tries to connect to a remote server.
As far as I understand this means that there is not operating system network configuration needed to achieve a transparent proxy.

I am working on a transparent on-device proxy solution using network level packet redirects written completely in Rust. So far my approach has been to modify the packets after capturing them with WinDivert to redirect the packets to the proxy port. 
That works with limitations - I am able to redirect the relevant packets to the proxy port, but I am unable to e.g. redirect the packages to the loopback interface. At the moment I overcome this issue by letting my proxy listen on all devices and redirect to the local IP rather than loopback.

I would love to get a better understanding of the architecture choices of the mitmproxy setup to improve the setup I am working on.

Here are some questions to kick this off:

* What is the main purpose of the virtual device in the mitmproxy setup and what are advantages (/disadvantages) of using the virtual device?
* My understanding is that you don't modify IP/TCP packages at all, but only re-create them when communicating them from the virtual device to the proxy, is that understanding correct?
* You introduced you own code repo for parsing network packages (i.e. https://docs.rs/internet-packet/latest/internet_packet/) even though you use smoltcp in other parts of the code already. Is there a specific reason why you use your own parse over smoltcp?
    * I am trying to parse the data with smoltcp but it has been a bit painful - I wonder if you hit any obstacles that I can avoid...
    * Do you have plans to extend https://docs.rs/internet-packet/latest/internet_packet/ to allow e.g. payload modification?

Thanks a lot in advance, I really love the work you are doing!

<details>
<summary> 👈 Unfold here to see a sequence diagram that sums up my understanding of the redirector setup in mitmproxy</summary>

This diagram shows the rough flow of the mitmproxy setup with the network level redirector. I am not totally sure my understanding is always correct and the diagram might not be complete. I added because it might be useful for the discussion in case I misunderstand the flow / concepts.
I am happy to provide the PlantUML code for this diagram if you think it is of value for the community.

![image](https://github.com/user-attachments/assets/c7780761-ad81-4f72-baaa-9be98e69b6b9)

</details>

**Comment by mhils** [comment]
> What is the main purpose of the virtual device in the mitmproxy setup and what are advantages (/disadvantages) of using the virtual device?

Are you talking about https://github.com/mitmproxy/mitmproxy_rs/blob/main/src/network/virtual_device.rs? This exists because we need something that implements https://docs.rs/smoltcp/latest/smoltcp/phy/trait.Device.html. From smoltcp's perspective, this implements the "outside world" from where it receives raw packets / where it transfers raw packets to. In mitmproxy's case this is then either forwarded to boringtun (WireGuard) or to WinDivert.

To clarify, this is no virtual TUN/TAP device or anything of that sort. This is just how we interface with smoltcp.

> My understanding is that you don't modify IP/TCP packages at all, but only re-create them when communicating them from the virtual device to the proxy, is that understanding correct?

Yes, that's pretty much correct.

> You introduced you own code repo for parsing network packages (i.e. https://docs.rs/internet-packet/latest/internet_packet/) even though you use smoltcp in other parts of the code already. Is there a specific reason why you use your own parse over smoltcp?

smoltcp didn't really have the right APIs for when this was built (e.g., IPv6 packets did not own their extension headers). internet-packet used to be part of mitmproxy_rs, but it's an external crate now because I wanted to re-use this for some work at my previous job. I think the smoltcp APIs have improved and matured quite a bit since then, so I can't offer any good advice. :)

> Do you have plans to extend https://docs.rs/internet-packet/latest/internet_packet/ to allow e.g. payload modification?

Nope, sorry! :)
